File: BitcoinCashExplorer_Bot.py
import requests
import json
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, BaseFilter
import telegram
import logging
updater = Updater(token="")
bot=telegram.Bot(token="")
dispatcher=updater.dispatcher
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.DEBUG)




def start(bot, update):
    bot.send_message(chat_id=update.message.chat_id, text="Hello, there is a bot for the BitcoinCash Blockchain Explorer. You can use these command:\n"
                     "/check_address\n"
                     "/check_transactions\n"
                     "/blockchainstatus\n"
                     "/price\n"
                     "/start (this message)\n"
                     "\n"
                     "<b>If you add this bot in a chat group, do not give him permission to read the messages</b>\n"
                     "\n"
                     "If you have a problem with bot or you want send me your feedback, contact me @MarckTomack\n", parse_mode=telegram.ParseMode.HTML)
start_handler=CommandHandler("start", start)
dispatcher.add_handler(start_handler)


def check_address(bot, update):
    bot.send_message(chat_id=update.message.chat_id, text="Send me the BitcoinCash address for check the balance")
check_a_handler=CommandHandler("check_address", check_address)
dispatcher.add_handler(check_a_handler)

def check_transactions(bot, update):
    bot.send_message(chat_id=update.message.chat_id, text="Send me the transaction hash for check the informations in the transaction")
check_t_handler=CommandHandler("check_transactions", check_transactions)
dispatcher.add_handler(check_t_handler)


class Filter_address(BaseFilter):
    def filter(self, message):
        return len(message.text) == 34 or len(message.text) == 42
filter_address= Filter_address()
def addr(bot,update):
    addr=(update.message.text)
    link="https://api.blockchair.com/bitcoin-cash/dashboards/address/"
    get_address=requests.get(link+addr)
    date=get_address.json()
    satoshi=0.00000001
    conv=date["data"][addr]["address"]["balance"]
    final=satoshi*float(conv)
    tbbch = str(final)
    tbusd = str(date["data"][addr]["address"]["balance_usd"])
    t = str(date["data"][addr]["address"]["transaction_count"])
    bot.send_message(chat_id=update.message.chat_id, text=f"<b>Total Balance BCH:</b> <code>{tbbch}</code>"  + "\n"
                     f"<b>Total Balance USD:</b> <code>{tbusd}</code>" + "\n"
                     f"<b>Transactions:</b> <code>{t}</code>", parse_mode=telegram.ParseMode.HTML)
addr_handler=MessageHandler(filter_address, addr)
dispatcher.add_handler(addr_handler)

class Filter_tx(BaseFilter):
    def filter(self, message):
        return len(message.text) == 64
filter_tx=Filter_tx()
def tx(bot, update):
    tx=(update.message.text)
    link="https://api.blockchair.com/bitcoin-cash/dashboards/transaction/"
    get_tx=requests.get(link+tx)
    date=get_tx.json()
    satoshi=0.00000001
    conv=date["data"][tx]["transaction"]["input_total"]
    final=satoshi*float(conv)
    conv_two=date["data"][tx]["transaction"]["output_total"]
    final_two=satoshi*float(conv_two)
    current_block=date["context"]["state"]
    init_block=date["data"][tx]["transaction"]["block_id"]
    conf=current_block - init_block
    tivbch = str(final)
    tivusd = str(date["data"][tx]["transaction"]["input_total_usd"])
    tovbch = str(final_two)
    tovusd = str(date["data"][tx]["transaction"]["output_total_usd"])
    sb = str(date["data"][tx]["transaction"]["size"])
    confi = str(conf)
    fees = str(date["data"][tx]["transaction"]["fee"])
    coinb = str(date["data"][tx]["transaction"]["is_coinbase"])
    bot.send_message(chat_id=update.message.chat_id, text=f"<b>Total Input Value BCH:</b> <code>{tivbch}</code>" + "\n"
                     f"<b>Total Input Value USD:</b> <code>{tivusd}</code>" + "\n"
                     f"<b>Total Output Value BCH:</b> <code>{tovbch}</code>" + "\n"
                     f"<b>Total Output Valie USD:</b> <code>{tovusd}</code>" + "\n"
                     f"<b>Size (byte):</b> <code>{sb}</code>" + "\n"
                     f"<b>Confirmations:</b> <code>{confi}</code>" + "\n"
                     f"<b>Fees:</b> <code>{fees} satoshi</code>" + "\n"
                     f"<b>Coinbase:</b> <code>{coinb}</code>", parse_mode=telegram.ParseMode.HTML)
tx_handler=MessageHandler(filter_tx, tx)
dispatcher.add_handler(tx_handler)




def blockchainstatus(bot, update):
    status="https://api.blockchair.com/bitcoin-cash/stats"
    status_get=requests.get(status)
    date=status_get.json()
    b = str(date["data"]["blocks"])
    bh = str(date["data"]["blocks_24h"])
    th = str("{:,}".format(date["data"]["transactions_24h"]))
    md = str("{:,}".format(date["data"]["difficulty"]))
    hr = str(date["data"]["hashrate_24h"])
    t = str("{:,}".format(date["data"]["transactions"]))
    o = str("{:,}".format(date["data"]["outputs"]))
    mt = str(date["data"]["mempool_transactions"])
    tfimusd = str(date["data"]["mempool_total_fee_usd"])
    tpsm = str(round(date["data"]["mempool_tps"], 2))
    cn = str("{:,}".format(date["data"]["nodes"]))
    cs = str("{:,}".format(date["data"]["circulation"]))
    bot.send_message(chat_id=update.message.chat_id, text=f"<b>Blocks:</b> <code>{b}</code>" + "\n"
                     f"<b>Blocks in 24h:</b> <code>{bh}</code>" + "\n"
                     f"<b>Transactions in 24h:</b> <code>{th}</code>" + "\n"
                     f"<b>Mining Difficulty:</b> <code>{md}</code>" + "\n"
                     f"<b>Hashrate 24h:</b> <code>{hr}</code>" + "\n"
                     f"<b>Transactions:</b> <code>{t}</code>" + "\n"
                     f"<b>Outputs:</b> <code>{o}</code>" + "\n"
                     f"<b>Mempool transactions:</b> <code>{mt}</code>" + "\n"
                     f"<b>Total fees in mempool (USD):</b> <code>{tfimusd}</code>" + "\n"
                     f"<b>Transactions per second in mempool:</b> <code>{tpsm}</code>" + "\n"
                     f"<b>Currently nodes:</b> <code>{cn}</code>" + "\n"
                     f"<b>Circulation supply:</b> <code>{cs}/21,000,000</code>", parse_mode=telegram.ParseMode.HTML)
status_handler=CommandHandler("blockchainstatus", blockchainstatus)
dispatcher.add_handler(status_handler)

def price(bot, update):
    link="https://api.blockchair.com/bitcoin-cash/stats"
    link_get=requests.get(link)
    date=link_get.json()
    p = str(round(date["data"]["market_price_usd"], 2))
    vh = str("{:,}".format(date["data"]["volume_24h"]))
    pch = str(round(date["data"]["market_price_usd_change_24h_percentage"], 2))
    mcusd = str("{:,}".format(date["data"]["market_cap_usd"]))
    qu = str(date["data"]["market_dominance_percentage"])
    bot.send_message(chat_id=update.message.chat_id, text=f"<b>Price USD:</b> <code>{p}</code>" + "\n"
                     f"<b>Volume 24h:</b> <code>{vh}</code>" + "\n"
                     f"<b>Percent change 24h:</b> <code>{pch}%</code>" + "\n"
                     f"<b>Market cap USD:</b> <code>{mcusd}</code>" + "\n"
                     f"<b>Market Dominance: </b> <code>{qu}%</code>", parse_mode=telegram.ParseMode.HTML)
price_handler=CommandHandler("price", price)
dispatcher.add_handler(price_handler)



#Error section

class Filter_address_no(BaseFilter):
    def filter(self, message):
        return len(message.text) != 34, 42, 64
filter_address_no=Filter_address_no()
def addrno(bot, update):
    bot.send_message(chat_id=update.message.chat_id, text="This is not a valid format of BitcoinCash address or a transaction hash, check if it's correct.\n"
                    "If your address starts with \"bitcoincash:\", remove it.")
no_handler=MessageHandler(filter_address_no, addrno)
dispatcher.add_handler(no_handler)



updater.start_polling()
updater.idle()

File: BitcoinCashExplorerBot.go
/* Copyright (C) MarckTomack <marcktomack@tutanota.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>. */

package main

import (
	"fmt"
	"github.com/buger/jsonparser"
	tb "gopkg.in/tucnak/telebot.v2"
	"io/ioutil"
	"math"
	"net/http"
	"strconv"
	"time"
)

const satoshi = 0.00000001

var (
	bot, _ = tb.NewBot(tb.Settings{
		Token:  "",
		Poller: &tb.LongPoller{Timeout: 10 * time.Second},
	})
)

func Start() {
	bot.Handle("/start", func(m *tb.Message) {
		bot.Send(m.Sender, "There is a bot for the BitcoinCash Blockchain Explorer."+"\n"+
			"You can use this command:"+"\n"+"\n"+
			"/blockchainstatus"+"\n"+
			"/price"+"\n"+"\n"+
			"Or send an address or a transaction hash for check informations about it."+"\n"+"\n"+
			"If you have a problem with the bot or you want send me your feedback, contact me @MarckTomack")
	})
}

func AddressAndTransaction() {
	bot.Handle(tb.OnText, func(m *tb.Message) {
		addr := m.Text
		tx := m.Text
		if len(addr) == 34 || len(addr) == 42 || len(addr) == 54 {
			url := "https://api.blockchair.com/bitcoin-cash/dashboards/address/"
			getRequest, _ := http.Get(url + addr)
			readBody, _ := ioutil.ReadAll(getRequest.Body)
			json := []byte(readBody)
			parseBalanceBCH, _ := jsonparser.GetFloat(json, "data", fmt.Sprintf(`%v`, addr), "address", "balance")
			balanceBCH := parseBalanceBCH * satoshi
			roundBalanceBCH := math.Round(balanceBCH)
			balanceBCHToPrint := fmt.Sprintf("%.2f", roundBalanceBCH)
			balanceUSD, _ := jsonparser.GetFloat(json, "data", fmt.Sprintf(`%v`, addr), "address", "balance_usd")
			roundBalanceUSD := math.Round(balanceUSD)
			balanceUSDToPrint := fmt.Sprintf("%.2f", roundBalanceUSD)

			transactionCount, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, addr), "address", "transaction_count")

			bot.Send(m.Sender, "<b>Total Balance BCH: </b>"+fmt.Sprintf(`<code>%v</code>`, balanceBCHToPrint)+"\n"+
				"<b>Total Balance USD: </b>"+
				fmt.Sprintf(`<code>%v</code>`, balanceUSDToPrint)+"\n"+
				"<b>Transactions: </b>"+fmt.Sprintf(`<code>%v</code>`, transactionCount), tb.ModeHTML)
		} else if len(tx) == 64 {
			url := "https://api.blockchair.com/bitcoin-cash/dashboards/transaction/"
			getRequest, _ := http.Get(url + tx)
			readBody, _ := ioutil.ReadAll(getRequest.Body)
			json := []byte(readBody)

			totalInputValueBCH, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "input_total")
			totalInputValueUSD, _ := jsonparser.GetFloat(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "input_total_usd")
			totalOutputValueBCH, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "output_total")
			totalOutputValueUSD, _ := jsonparser.GetFloat(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "output_total_usd")
			size, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "size")
			initBlock, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "block_id")
			currentBlock, _ := jsonparser.GetInt(json, "context", "state")
			confirmations := currentBlock - initBlock
			fees, _ := jsonparser.GetInt(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "fee")
			coinbase, _ := jsonparser.GetFloat(json, "data", fmt.Sprintf(`%v`, tx), "transaction", "is_coinbase")

			bot.Send(m.Sender, "<b>Total Input Value BCH: </b>"+fmt.Sprintf(`<code>%v</code>`, float64(totalInputValueBCH)*satoshi)+"\n"+
				"<b>Total Input Value USD: </b>"+fmt.Sprintf(`<code>%v</code>`, totalInputValueUSD)+"\n"+
				"<b>Total Output Value BCH: </b>"+fmt.Sprintf(`<code>%v</code>`, float64(totalOutputValueBCH)*satoshi)+"\n"+
				"<b>Total Output Value USD: </b>"+fmt.Sprintf(`<code>%v</code>`, totalOutputValueUSD)+"\n"+
				"<b>Size: </b>"+fmt.Sprintf(`<code>%v</code>`, size)+"\n"+
				"<b>Confirmations: </b>"+fmt.Sprintf(`<code>%v</code>`, confirmations)+"\n"+
				"<b>Fees: </b>"+fmt.Sprintf(`<code>%v</code>`, fees)+"\n"+
				"<b>Is Coinbase: </b>"+fmt.Sprintf(`<code>%v</code>`, coinbase), tb.ModeHTML)
		} else {
			bot.Send(m.Sender, "This is not a valid BCH address or transaction hash")
		}

	})
}

func BlockchainStatus() {
	bot.Handle("/blockchainstatus", func(m *tb.Message) {
		getRequest, _ := http.Get("https://api.blockchair.com/bitcoin-cash/stats")
		readBody, _ := ioutil.ReadAll(getRequest.Body)
		json := []byte(readBody)

		blocks, _ := jsonparser.GetInt(json, "data", "blocks")
		blocksIn24h, _ := jsonparser.GetInt(json, "data", "blocks_24h")
		transactions, _ := jsonparser.GetUnsafeString(json, "data", "transactions")
		transactionsIn24h, _ := jsonparser.GetInt(json, "data", "transactions_24h")
		miningDifficulty, _ := jsonparser.GetUnsafeString(json, "data", "difficulty")
		hashrate24h, _ := jsonparser.GetUnsafeString(json, "data", "hashrate_24h")
		outputs, _ := jsonparser.GetInt(json, "data", "outputs")
		mempoolTransactions, _ := jsonparser.GetInt(json, "data", "mempool_transactions")
		totalFeesInMempool, _ := jsonparser.GetFloat(json, "data", "mempool_total_fee_usd")
		roundTotalFeesInMempool := math.Round(totalFeesInMempool)
		totalFeesInMempoolToPrint := fmt.Sprintf("%.2f", roundTotalFeesInMempool)
		txPerSecondInMempool, _ := jsonparser.GetFloat(json, "data", "mempool_tps")
		roundTxPerSecondInMempool := math.Round(txPerSecondInMempool)
		txPerSecondInMempoolToPrint := fmt.Sprintf("%.2f", roundTxPerSecondInMempool)

		currentlyNodes, _ := jsonparser.GetInt(json, "data", "nodes")
		circulationSupply, _ := jsonparser.GetFloat(json, "data", "circulation")
		roundCirculationSupply := math.Round(circulationSupply)
		formatedCirculationSupply := fmt.Sprintf("%.2f", roundCirculationSupply*satoshi)
		circulationSupplyToPrint := formatedCirculationSupply + "/" + strconv.Itoa(21000000)

		bot.Send(m.Sender, "<b>Blocks: </b>"+fmt.Sprintf(`<code>%v</code>`, blocks)+"\n"+
			"<b>Blocks in 24h: </b>"+fmt.Sprintf(`<code>%v</code>`, blocksIn24h)+"\n"+
			"<b>Transactions: </b>"+fmt.Sprintf(`<code>%v</code>`, transactions)+"\n"+
			"<b>Transactions in 24h: </b>"+fmt.Sprintf(`<code>%v</code>`, transactionsIn24h)+"\n"+
			"<b>Mining Difficulty: </b>"+fmt.Sprintf(`<code>%v</code>`, miningDifficulty)+"\n"+
			"<b>Hashrate 24: </b>"+fmt.Sprintf(`<code>%v</code>`, hashrate24h)+"\n"+
			"<b>Outputs: </b>"+fmt.Sprintf(`<code>%v</code>`, outputs)+"\n"+
			"<b>Mempoll Transactions: </b>"+fmt.Sprintf(`<code>%v</code>`, mempoolTransactions)+"\n"+
			"<b>Total Fees in Mempool USD: </b>"+fmt.Sprintf(`<code>%v</code>`, totalFeesInMempoolToPrint)+"\n"+
			"<b>Transactions per Second in Mempool: </b>"+fmt.Sprintf(`<code>%v</code>`, txPerSecondInMempoolToPrint)+"\n"+
			"<b>Currently Nodes: </b>"+fmt.Sprintf(`<code>%v</code>`, currentlyNodes)+"\n"+
			"<b>Circulation Supply: </b>"+fmt.Sprintf(`<code>%v</code>`, circulationSupplyToPrint), tb.ModeHTML)
	})
}

func Price() {
	bot.Handle("/price", func(m *tb.Message) {
		getRequest, _ := http.Get("https://api.blockchair.com/bitcoin-cash/stats")
		readBody, _ := ioutil.ReadAll(getRequest.Body)
		json := []byte(readBody)

		priceUSD, _ := jsonparser.GetFloat(json, "data", "market_price_usd")
		volume24h, _ := jsonparser.GetUnsafeString(json, "data", "volume_24h")
		percentChange24h, _ := jsonparser.GetFloat(json, "data", "market_price_usd_change_24h_percentage")
		roundPercentChange := math.Round(percentChange24h)
		formatedPercentChange := fmt.Sprintf("%.2f", roundPercentChange)
		percentChangeToPrint := formatedPercentChange + "%"
		marketcapUSD, _ := jsonparser.GetUnsafeString(json, "data", "market_cap_usd")
		marketDominance, _ := jsonparser.GetFloat(json, "data", "marke_dominance_percentage")
		marketDominanceToPrint := strconv.Itoa(int(marketDominance)) + "%"

		bot.Send(m.Sender, "<b>Price USD: </b>"+fmt.Sprintf(`<code>%v</code>`, priceUSD)+"\n"+
			"<b>Volume 24h: </b>"+fmt.Sprintf(`<code>%v</code>`, volume24h)+"\n"+
			"<b>Percent Change 24h: </b>"+fmt.Sprintf(`<code>%v</code>`, percentChangeToPrint)+"\n"+
			"<b>Market Cap USD: </b>"+fmt.Sprintf(`<code>%v</code>`, marketcapUSD)+"\n"+
			"<b>Market Dominance: </b>"+fmt.Sprintf(`<code>%v</code>`, marketDominanceToPrint), tb.ModeHTML)

	})

}

func main() {
	Start()
	AddressAndTransaction()
	BlockchainStatus()
	Price()
	bot.Start()
}
